File: generator.cpp
/**
 * MeterFeed Library
 * 
 * by fp2.dev
 */

#include "generator.h"

MeterFeeder::Generator::Generator(char* serialNumber, char* description, FT_HANDLE handle) {
    serialNumber_ = serialNumber;
    description_ = description;
    ftHandle_ = handle;
};

char* MeterFeeder::Generator::GetSerialNumber() {
    return serialNumber_;
};

char* MeterFeeder::Generator::GetDescription() {
    return description_;
};

FT_HANDLE MeterFeeder::Generator::GetHandle() {
    return ftHandle_;
};

int MeterFeeder::Generator::Stream() {
    UCHAR initCommand = 0x96; // Streaming command
    DWORD bytesTxd = 0;

	// Purge before writing
	FT_STATUS ftdiStatus = FT_Purge(ftHandle_, FT_PURGE_RX | FT_PURGE_TX);
    if (ftdiStatus != FT_OK) {
		return MF_DEVICE_ERROR;
	}

    // WRITE TO DEVICE
	ftdiStatus = FT_Write(ftHandle_, &initCommand, 1, &bytesTxd);
	if (ftdiStatus != FT_OK || bytesTxd != 1) {
		return MF_DEVICE_ERROR;
	}

    return MF_OK;
}

int MeterFeeder::Generator::Read(UCHAR* dxData) {
	DWORD bytesRxd = 0;

    // READ FROM DEVICE
	FT_STATUS ftdiStatus = FT_Read(ftHandle_, dxData, 1, &bytesRxd);
	if (ftdiStatus != FT_OK || bytesRxd != 1) {
		return MF_DEVICE_ERROR;
	}

    return MF_OK;
}

void MeterFeeder::Generator::Close() {
    FT_Close(ftHandle_);
}

File: android/MeterFeeder/MeterFeeder/src/main/java/dev/fp2/meterfeeder/Generator.java
package dev.fp2.meterfeeder;

import com.ftdi.j2xx.D2xxManager;
import com.ftdi.j2xx.FT_Device;

public class Generator {
    final static byte FT_PURGE_RX = 1;
    final static byte FT_PURGE_TX = 2;
    final static byte FTDI_DEVICE_START_STREAMING_COMMAND = (byte)0x96;

    D2xxManager ftdid2xx_;
    FT_Device device_;
    String serialNumber_;
    String description_;
    int ftHandle_;

    public Generator(D2xxManager ftdid2xx, FT_Device device, String serialNumber, String description, int handle) {
        ftdid2xx_ = ftdid2xx;
        device_ = device;
        serialNumber_ = serialNumber;
        description_ = description;
        ftHandle_ = handle;
    };

    public String GetSerialNumber() {
        return serialNumber_;
    };

    public String GetDescription() {
        return description_;
    };

    public int GetHandle() {
        return ftHandle_;
    };

    public boolean Stream() {
        // Purge before writing
        if (!device_.purge((byte)(FT_PURGE_RX | FT_PURGE_TX))) {
            return false;
        }

        // WRITE TO DEVICE
        if (device_.write(new byte[] {FTDI_DEVICE_START_STREAMING_COMMAND}, 1, true) != 1) {
            return false;
        }

        return true;
    }

    public byte[] Read(int length) {
        byte[] entropyBytes = new byte[length];

        // READ FROM DEVICE
        int read = device_.read(entropyBytes, length);
        if (read != length) {
            return null;
        }

        return entropyBytes;
    }

    public void Close() {
        device_.close();
    }
}
